ddtrace/contrib/pytest/_efd_utils.py

```diff
@@ -0,0 +1,243 @@
+from dataclasses import dataclass
+import typing as t
+
+from _pytest.logging import caplog_handler_key
+from _pytest.logging import caplog_records_key
+import pytest
+
+from ddtrace.contrib.pytest._retry_utils import RetryOutcomes
+from ddtrace.contrib.pytest._retry_utils import _efd_get_attempt_string
+from ddtrace.contrib.pytest._retry_utils import _retry_run_when
+from ddtrace.contrib.pytest._retry_utils import set_retry_num
+from ddtrace.contrib.pytest._utils import _get_test_id_from_item
+from ddtrace.contrib.pytest._utils import _TestOutcome
+from ddtrace.ext.test_visibility.api import TestExcInfo
+from ddtrace.ext.test_visibility.api import TestStatus
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.test_visibility._efd_mixins import EFDTestStatus
+from ddtrace.internal.test_visibility.api import InternalTest
+
+
+log = get_logger(__name__)
+
+
+@dataclass(frozen=True)
+class __EFD_RETRY_OUTCOMES:
+    EFD_ATTEMPT_PASSED = "dd_efd_attempt_passed"
+    EFD_ATTEMPT_FAILED = "dd_efd_attempt_failed"
+    EFD_ATTEMPT_SKIPPED = "dd_efd_attempt_skipped"
+    EFD_FINAL_PASSED = "dd_efd_final_passed"
+    EFD_FINAL_FAILED = "dd_efd_final_failed"
+    EFD_FINAL_SKIPPED = "dd_efd_final_skipped"
+    EFD_FINAL_FLAKY = "dd_efd_final_flaky"
+
+
+_EFD_RETRY_OUTCOMES = __EFD_RETRY_OUTCOMES()
+_EFD_FLAKY_OUTCOME = "Flaky"
+
+_FINAL_OUTCOMES: t.Dict[EFDTestStatus, str] = {
+    EFDTestStatus.ALL_PASS: _EFD_RETRY_OUTCOMES.EFD_FINAL_PASSED,
+    EFDTestStatus.ALL_FAIL: _EFD_RETRY_OUTCOMES.EFD_FINAL_FAILED,
+    EFDTestStatus.ALL_SKIP: _EFD_RETRY_OUTCOMES.EFD_FINAL_SKIPPED,
+    EFDTestStatus.FLAKY: _EFD_RETRY_OUTCOMES.EFD_FINAL_FLAKY,
+}
+
+
+def efd_handle_retries(test_id, item, when, original_result, test_outcome):
+    # Don't mark the original test's call as failed if it will be EFD-retried to avoid failing the session
+    # pytest_terminal_summary will then update the counts according to the test's final status
+    if when == "call" and test_outcome.status == TestStatus.FAIL:
+        original_result.outcome = _EFD_RETRY_OUTCOMES.EFD_ATTEMPT_FAILED
+        return
+    if InternalTest.get_tag(test_id, "_dd.ci.efd_setup_failed"):
+        log.debug("Test item %s failed during setup, will not be retried for Early Flake Detection")
+        return
+    if InternalTest.get_tag(test_id, "_dd.ci.efd_teardown_failed"):
+        # NOTE: tests that passed their call but failed during teardown are not retried
+        log.debug("Test item %s failed during teardown, will not be retried for Early Flake Detection")
+        return
+
+    original_result.outcome = _EFD_RETRY_OUTCOMES.EFD_ATTEMPT_PASSED
+    efd_outcome = _efd_handle_retries(item, test_outcome)
```

Comment:
Having two functions named `efd_handle_retries` and `_efd_handle_retries` might be somewhat confusing, maybe we could name `_efd_handle_retries` something more specific.